frontends/systolic-lang/gen-systolic.py

```diff
@@ -323,52 +420,58 @@ def counter():
 
     # end source pos init
 
-    # Increment memories for PE_00 before computing with it.
-    upd_pe00_mem = []
-    upd_pe00_mem.append(py_ast.Enable(NAME_SCHEME["index update"].format(prefix="t0")))
-    upd_pe00_mem.append(py_ast.Enable(NAME_SCHEME["index update"].format(prefix="l0")))
-    control.append(py_ast.ParComp(upd_pe00_mem))
-
-    for (idx, elements) in enumerate(sch):
-        # Move all the requisite data.
-        move: list[py_ast.Control] = [
-            py_ast.Enable(row_data_mover_at(r, c)) for (r, c) in elements
-        ]
-        move.extend([py_ast.Enable(col_data_mover_at(r, c)) for (r, c) in elements])
-        control.append(py_ast.ParComp(move))
-
-        # Update the indices if needed.
-        more_control = []
-        if idx < len(sch) - 1:
-            next_elements = sch[idx + 1]
-            upd_memory = [
-                py_ast.Enable(upd)
-                for (r, c) in next_elements
-                if (r == 0 or c == 0)
-                for upd in index_update_at(r, c)
-            ]
-            more_control.extend(upd_memory)
-
-        # py_ast.Invoke the PEs and move the data to the next layer.
-        for (r, c) in elements:
-
-            invoke = py_ast.Invoke(
-                id=py_ast.CompVar(f"pe_{r}_{c}"),
-                in_connects=[
-                    ("top", py_ast.CompPort(py_ast.CompVar(f"top_{r}_{c}"), "out")),
-                    (
-                        "left",
-                        py_ast.CompPort(py_ast.CompVar(f"left_{r}_{c}"), "out"),
-                    ),
-                ],
-                out_connects=[],
+    control_stmts = []
+    incr_stmts = [py_ast.Enable("incr_idx")]
+    for r in range(left_length):
+        for c in range(top_length):
+            # build 4 if stmts for the 4 schedules that we need to account for
+            input_mem_updates = execute_if_between(
```

Comment:
This is cool! It's like we're building a DSL of "instructions" (like `execute_if_between`) which then gets turned into hardware. You can imagine that in the future, we could build a systolic array DSL with literally these operations and have users program in it!